src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/services/RestrictedPatientsService.kt

```diff
@@ -73,6 +75,49 @@ class RestrictedPatientsService(
       .let { addRestrictedPatient(it) }
   }
 
+  @Transactional
+  fun prisonerReleased(offenderNo: String) {
+    if (restrictedPatientsRepository.existsById(offenderNo)) {
+      log.info("Ignoring release of prisoner $offenderNo as they are already a restricted patient")
+      return
+    }
+
+    val lastMovement = getReleaseToHospitalMovement(offenderNo)
+    if (lastMovement == null) {
+      log.info("Ignoring release of prisoner $offenderNo as they are not being released to a hospital")
+      return
+    }
+
+    val hospital = getHospital(lastMovement)
+    if (hospital == null) {
+      log.info("Ignoring release of prisoner $offenderNo due to unrecognised hospital ${lastMovement.toAgency}")
+      return
+    }
+
+    restrictedPatientsRepository.save(
+      RestrictedPatient(
+        prisonerNumber = offenderNo,
+        fromLocationId = lastMovement.fromAgency!!,
+        hospitalLocationCode = hospital.agencyId,
+        supportingPrisonId = lastMovement.fromAgency,
+        dischargeTime = calculateDischargeDateTime(offenderNo, lastMovement.movementDate, lastMovement.movementTime),
+        commentText = lastMovement.commentText,
+      ),
+    )
+    domainEventPublisher.publishRestrictedPatientAdded(offenderNo)
+  }
+
+  private fun getHospital(lastMovement: MovementResponse): Agency? =
+    lastMovement
+      .takeIf { it.toAgency != null }
+      ?.let { prisonApiGateway.getAgency(it.toAgency!!) }
+      ?.takeIf { agency -> listOf("HOSPITAL", "HSHOSP").contains(agency.agencyType) }
+
+  private fun getReleaseToHospitalMovement(offenderNo: String): MovementResponse? =
+    prisonApiGateway.getLatestMovements(offenderNo)
```

Comment:
could you call `getLatestMovement` in this class rather than doing this

Reply:
I could, but then I've got to catch the BadRequestException because I don't want to retry the message. Also, that check for `movement.size > 1` is superfluous. So that's why I didn't.